spacy/pipeline/transition_parser.pyx

```diff
@@ -757,3 +774,46 @@ def states2actions(states: List[StateClass]) -> List[Ints1d]:
         actions.append(numpy.array(step_actions, dtype="i"))
 
     return actions
+
+def _states_diff_to_actions(
+    before_states: List[StateClass],
+    after_states: List[StateClass]
+) -> List[Ints1d]:
+    """
+    Return for two sets of states the actions to go from the first set of
+    states to the second set of states. The histories of the first set of
+    states must be a prefix of the second set of states.
+    """
+    cdef StateClass before_state, after_state
+    cdef StateC* c_state_before
+    cdef StateC* c_state_after
+
+    assert len(before_states) == len(after_states)
+
+    # Check invariant: before states histories must be prefixes of after states.
+    for before_state, after_state in zip(before_states, after_states):
+        c_state_before = before_state.c
+        c_state_after = after_state.c
+
+        assert equal(c_state_after.history.begin(),
+            c_state_after.history.begin() + c_state_before.history.size(),
+            c_state_after.history.begin())
```

Comment:
That should be `c_state_before.history.begin()`, right? 

I also wonder if we should assert the lengths of the before and after histories before we do any kind of pointer/iterator arithmetic on them. Maybe I'm just being paranoid, but we can't guarantee that `c_state_after.history.begin() + c_state_before.history.size()` won't trigger a segfault when it's dereferenced, e.g: when `c_state_after.history` is smaller than `c_state_before.history` and is weirdly placed at a boundary bordering a page that doesn't have the read permissions.